### ISSUE TITLE
Building variational auto-encoder

### ISSUE THREAD

**Issue by tianxiangchen2015** [NervanaSystems/neon#363]
Hi,

I'm trying to build a variational auto-encoder. I want to create a layer that sampling the inputs from previous layer. Here's my implementation.

`from neon.layers.layer import Layer, interpret_in_shape
class Lambda(Layer):

    # constructor and initialize buffers
    def __init__(self, nout, name=None):
        super(Lambda, self).__init__(name)
        self.nout = nout

    # configure the layer input and output shapes
    def configure(self, in_obj):
        super(Lambda, self).configure(in_obj)
        (self.nin, self.nsteps) = interpret_in_shape(self.in_shape)
        self.out_shape = (self.nout, self.nsteps)
        return self

    # compute the fprop
    def fprop(self, inputs, inference=False):
        self.inputs = inputs
        
        def sampling(data):
            z_mean = z_log_var = data
            epsilon = np.random.normal(0, 1.0, (data.shape))
            return z_mean + np.exp(z_log_var)*epsilon
        
        for c in range(self.outputs.shape[1]):
            self.outputs[:,c] = sampling(self.inputs[:,c])
        return self.outputs

    # backprop the gradients
    def bprop(self, error):
        error[:] = error*1
        return error`

Here's my architecture.

layers = [Affine(nout=128, init=init, activation=activation),
                Affine(nout=3, init=init, activation=activation),
                Lambda(nout=3),
                Affine(nout=128, init=init, activation=activation),
                Affine(nout=28*28, init=init, activation=activation)]

After I added this `Lambda` layer in my model, it doesn't give me any error. But, the training process just can not start. Is there any thing wrong in my implementation?

Thanks

**Comment by yinyinl** [comment]
Can you tell us more what did you mean by "training process just can not start". Any error message? It doesn't seem that you can also just do `self.out_shape = self.in_shape`. Also, are you running on GPU? if you do, this part may not work given the data are on different memories `z_mean + np.exp(z_log_var)*epsilon`.
